builtin/logical/database/version_wrapper.go

```diff
@@ -0,0 +1,268 @@
+package database
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/helper/random"
+	"github.com/hashicorp/vault/sdk/database/dbplugin"
+	"github.com/hashicorp/vault/sdk/database/newdbplugin"
+	"github.com/hashicorp/vault/sdk/helper/pluginutil"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type databaseVersionWrapper struct {
+	v4 dbplugin.Database
+	v5 newdbplugin.Database
+}
+
+// newDatabaseWrapper figures out which version of the database the pluginName is referring to and returns a wrapper object
+// that can be used to make operations on the underlying database plugin.
+func newDatabaseWrapper(ctx context.Context, pluginName string, sys pluginutil.LookRunnerUtil, logger log.Logger) (dbw databaseVersionWrapper, err error) {
+	newDB, err := newdbplugin.PluginFactory(ctx, pluginName, sys, logger)
+	if err == nil {
+		dbw = databaseVersionWrapper{
+			v5: newDB,
+		}
+		return dbw, nil
+	}
+
+	legacyDB, err := dbplugin.PluginFactory(ctx, pluginName, sys, logger)
+	if err == nil {
+		dbw = databaseVersionWrapper{
+			v4: legacyDB,
+		}
+		return dbw, nil
+	}
+
+	return dbw, fmt.Errorf("invalid database version")
+}
+
+// Initialize the underlying database. This is analogous to a constructor on the database plugin object.
+// Errors if the wrapper does not contain an underlying database.
+func (d databaseVersionWrapper) Initialize(ctx context.Context, req newdbplugin.InitializeRequest) (newdbplugin.InitializeResponse, error) {
+	if !d.isNewDB() && !d.isLegacyDB() {
```

Comment:
`newDatabaseWrapper` should disallow this from happening right?

Reply:
Yes, but this is protecting against creating a `databaseVersionWrapper` instance directly without calling `newDatabaseWrapper`.

Reply:
Fair enough. I always go back and forth between nil checking or letting the constructor handle sane defaults and proper guarantees (more applicable around the receiver itself when it's a pointer value and not on its fields, but in similar spirit).